applications/external/picopass/picopass_worker.c

```diff
@@ -749,3 +787,484 @@ void picopass_worker_write_key(PicopassWorker* picopass_worker) {
+}
+
+void picopass_worker_emulate(PicopassWorker* picopass_worker, bool loclass_mode) {
+    FuriHalNfcTxRxContext tx_rx = {};
+    PicopassEmulatorCtx emu_ctx = {
+        .state = PicopassEmulatorStateIdle,
+        .key_block_num = PICOPASS_SECURE_KD_BLOCK_INDEX,
+        .loclass_mode = loclass_mode,
+        .loclass_got_std_key = false,
+        .loclass_writer = NULL,
+    };
+    FuriHalNfcDevData nfc_data = {
+        .uid_len = RFAL_PICOPASS_UID_LEN,
+    };
+    NfcVData* nfcv_data = malloc(sizeof(NfcVData));
+    nfcv_data->block_size = RFAL_PICOPASS_BLOCK_LEN;
+    nfcv_data->emu_protocol_ctx = &emu_ctx;
+    nfcv_data->emu_protocol_handler = &picopass_emu_handle_packet;
+
+    PicopassDeviceData* dev_data = picopass_worker->dev_data;
+    PicopassBlock* blocks = dev_data->AA1;
+
+    if(loclass_mode) {
+        // Setup blocks for loclass attack
+        emu_ctx.key_block_num = 0;
+        loclass_update_csn(&nfc_data, nfcv_data, &emu_ctx);
+
+        uint8_t conf[8] = {0x12, 0xFF, 0xFF, 0xFF, 0x7F, 0x1F, 0xFF, 0x3C};
+        picopass_emu_write_blocks(nfcv_data, conf, PICOPASS_CONFIG_BLOCK_INDEX, 1);
+
+        uint8_t epurse[8] = {0xFE, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF};
+        picopass_emu_write_blocks(nfcv_data, epurse, PICOPASS_SECURE_EPURSE_BLOCK_INDEX, 1);
+
+        uint8_t aia[8] = {0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF};
+        picopass_emu_write_blocks(nfcv_data, aia, PICOPASS_SECURE_AIA_BLOCK_INDEX, 1);
+
+        emu_ctx.loclass_writer = loclass_writer_alloc();
+        loclass_writer_write_start_stop(emu_ctx.loclass_writer, true);
+    } else {
+        memcpy(nfc_data.uid, blocks[PICOPASS_CSN_BLOCK_INDEX].data, RFAL_PICOPASS_BLOCK_LEN);
+        memcpy(nfcv_data->data, blocks, sizeof(dev_data->AA1));
+        picopass_init_cipher_state(nfcv_data, &emu_ctx);
+    }
+
+    uint8_t last_loclass_csn_num = 0;
+    bool loclass_got_std_key = false;
+
+    nfcv_emu_init(&nfc_data, nfcv_data);
+    while(picopass_worker->state == PicopassWorkerStateEmulate ||
+          picopass_worker->state == PicopassWorkerStateLoclass) {
+        if(nfcv_emu_loop(&tx_rx, &nfc_data, nfcv_data, 500)) {
+            if(picopass_worker->callback) {
+                if((loclass_mode) && (last_loclass_csn_num != emu_ctx.key_block_num)) {
+                    last_loclass_csn_num = emu_ctx.key_block_num;
+                    picopass_worker->callback(
+                        PicopassWorkerEventLoclassGotMac, picopass_worker->context);
+                } else if((loclass_mode) && !loclass_got_std_key && emu_ctx.loclass_got_std_key) {
+                    loclass_got_std_key = true;
+                    picopass_worker->callback(
+                        PicopassWorkerEventLoclassGotStandardKey, picopass_worker->context);
+                } else {
+                    picopass_worker->callback(
+                        PicopassWorkerEventSuccess, picopass_worker->context);
```

Comment:
In the other similar blocks they'll have a "break" after something like this.  I don't fully understand the emulation code well enough to say we should have one, but I just wanted to point it out in case it was missed.